### ISSUE TITLE
Make the Zoom API Url configurable in the Connection settings

### ISSUE THREAD

**Issue by didier63** [ncstate-delta/moodle-mod_zoom#261]
Hi,
For GDPR reason, we would like to connect the plugin to European Zoom servers.
https://eu01api-www4local.zoom.us for example.
Is it possible to make this parameter configurable ?
Kind regards,
Didier

**Comment by jrchamp** [comment]
### Background:
The Zoom API base URL is set here:
https://github.com/ncstate-delta/moodle-mod_zoom/blob/09562dcceadcc28b533e36138c05262c783b392b/classes/webservice.php#L58

The documentation describes only two possible options at this time: https://marketplace.zoom.us/docs/api-reference/introduction

### Thoughts:
1. With only two options, a select list makes a lot of sense to limit the possibility of a mistake.
2. Alternatively, a free text field with a good default and a description / link to the documentation would be more forward-compatible, but may lead to a credential leak if a malicious user is simply able to overwrite the configuration value (admittedly, this would require a vulnerability somewhere else in the Moodle code).

**Comment by abias** [comment]
This is an interesting and important discussion.

We are a EU customer and have a EU provisioning contract. However, I was not aware that there is a separate URL for handlung API calls for EU customers. Naively, I had somehow expected that Zoom handles the global API communication based on the JWT credentials we are passing to the Zoom API.

And looking at https://marketplace.zoom.us/docs/api-reference/introduction today, I do not see any note about `eu01api-www4local.zoom.us` at all.

@didier63 or @jrchamp , can you point me to the official documentation (URL or screenshot) where this `eu01api-www4local.zoom.us` domain is mentioned?

@didier63 , if you change the constant which @jrchamp mentioned to hold `eu01api-www4local.zoom.us` instead of `api.zoom.us`, does the Zoom plugin still work expected?

Furthermore, regarding this issue here, if there are really just two options (one for global customers and one for EU provisioning), then I would strongly encourage you to build a select list with a clear explanation which option to pick when.

As soon as the way forward is agreed, I will be happy to support in implementing and testing a patch.

Cheers,
Alex

**Comment by jrchamp** [comment]
Yeah, the next time I checked it was gone. I don't know what happened there. It seems like Zoom should just make their standard API endpoint EU-compatible and maybe that's what the did? I have not idea.

**Comment by abias** [comment]
> It seems like Zoom should just make their standard API endpoint EU-compatible and maybe that's what the did?

At Ulm university, we have asked the Zoom support (or at least our local Zoom reseller) to tell us which URL to use to be EU-compatible and what happens if we have a EU deployment but communicate with the Zoom API on api.zoom.us.

I will post the answer here as soon as I have got it.

**Comment by abias** [comment]
We have finally received some feedback from our local Zoom reseller now. These are the results:

* The API endpoint `eu01api-www4local.zoom.us` is an official API endpoint for Zoom customers who have a Zoom license with EU provisioning.
* The API endpoint `eu01api-www4local.zoom.us` currently resolves to servers in Frankfurt, Germany, while the default API endpoint `api.zoom.us` resolves to servers in the USA. Thus, at least on the first hop, the data stays in Europe.
* Zoom customers who have a Zoom license with EU provisioning are – technically – able to use the API endpoint `api.zoom.us` as well. Using `eu01api-www4local.zoom.us` is not a necessity for them.
* The API endpoint `eu01api-www4local.zoom.us` was removed from the docs page at https://marketplace.zoom.us/docs/api-reference/introduction as the explanation regarding the GDPR implications were somehow misleading. Our local Zoom reseller has asked Zoom to rephrase the explanation and publish the information officially again.

Having said that, I think that it would be wise to have an admin setting to let the admin decide between the global API endpoint and the EU API endpoint. I will contribute a PR for this setting soon.
